### ISSUE TITLE
Vega triggers override mistake when overriding O.p.hasOwnProperty by assignment

### ISSUE THREAD

**Issue by erights** [vega/vega#3075]
At docs/vega-core.js line 32463 and at doc/vega.js line 45238 appears the line

```js
  exports.hasOwnProperty = has;
```

This `exports` object inherits from `Object.prototype` which has a method named `hasOwnProperty`. Unfortunately, because of the [EcmaScript Override Mistake](https://web.archive.org/web/20141230041441/http://wiki.ecmascript.org/doku.php?id=strawman:fixing_override_mistake) (also [here](https://github.com/tc39/ecma262/pull/1307) and [here](https://github.com/tc39/ecma262/pull/1320)), in running in an environment (like SES, XS, or TC53 JavaScript for Embedded Systems) where `Object.prototype` is frozen, the Override Mistake will cause that assignment to fail.

From the line numbers above, it is clear these two files are generated files. However I do not understand what the actual source lines are that are translated into these generated lines, so I'm not sure what the best fix is. Two immediate possibilities: 
   * Have `exports` not inherit from `Object.prototype`. Depending on the role of `exports` in the translation, I would not be surprised if that is not feasible.
   * Use `defineProperty` rather than assignment, such as
      ```js
       Object.defineProperty(exports, 'hasOwnProperty', {
           value: has, 
           enumerable: true, 
           writable: true, 
           configurable: true 
       });
       ```

See https://github.com/Agoric/agoric-sdk/issues/2373 for the consequences of this bug in the Agoric use of SES.

**Comment by erights** [comment]
See https://github.com/Agoric/agoric-sdk/pull/2575

**Comment by domoritz** [comment]
I suspect the problem is in https://github.com/vega/vega/blob/master/packages/vega-util/src/hasOwnProperty.js. Do you have an idea how we can resolve the issue you are seeing?

**Comment by erights** [comment]
No, that's not the problem. There is nothing about that file that would trigger the override mistake. 

My first post above https://github.com/vega/vega/issues/3075#issue-805227532 explains the problem well, but complains about the wrong file. The problem is in vega-util/build/vega-util.js . Aside from that, please ask about anything unclear in the explanation in that first post.

Vega actually has three problems along these lines, for `constructor`, 'hasOwnProperty` and `toString`, all in the vega-util.js file. A patch which fixes all these is at
https://github.com/Agoric/agoric-sdk/blob/master/patches/vega-util%2B1.16.0.patch

**Comment by erights** [comment]
I just looked again at your sources rather than my patch. It looks like my patch is against, not a source file, but a file built from those sources. Is that correct? If so, then my first post was correct. The problem is in docs/vega-code.js and docs/vega.js .
However, both of these are *huge* files. Much larger than I'd expect from a source file. But the string `exports.hasOwnProperty` appears in your source tree in these two files and no where else. Where does this text come from?

**Comment by erights** [comment]
See https://github.com/vega/vega/pull/3109

**Comment by jheer** [comment]
Closed via #3109
